<a href="https://colab.research.google.com/github/kaixing189/MSSP607/blob/main/WeeklyModules/Week05/Week_05_Participation_Activity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MSSP607 Week 5 Participation Activity


## Part 1. Clone the GitHub Repository


In [13]:
from pathlib import Path
import os
import subprocess

REPO_URL = "https://github.com/kaixing189/MSSP607.git"
REPO_DIR = Path("/content/MSSP607")

if REPO_DIR.exists():
    subprocess.run(
        ["git", "-C", str(REPO_DIR), "pull"],
        check=True
    )
    print("Repository already exists. Latest changes were pulled.")
else:
    subprocess.run(
        ["git", "clone", REPO_URL, str(REPO_DIR)],
        check=True
    )
    print("Repository cloned successfully.")

os.chdir(REPO_DIR)
print("Current working directory:", os.getcwd())

Repository already exists. Latest changes were pulled.
Current working directory: /content/MSSP607


## Part 2. Define the Data Folder and Check the Files



In [14]:
DATA_DIR = REPO_DIR / "data"

required_files = [
    "aapl.csv",
    "amzn.csv",
    "fb.csv",
    "goog.csv",
    "nflx.csv"
]

for filename in required_files:
    file_path = DATA_DIR / filename
    print(f"{filename}: {file_path.exists()}")

assert all((DATA_DIR / filename).exists() for filename in required_files),     "One or more required CSV files are missing."

aapl.csv: True
amzn.csv: True
fb.csv: True
goog.csv: True
nflx.csv: True


## Part 3. Import pandas and Read Each CSV File



In [15]:
import pandas as pd

aapl = pd.read_csv(DATA_DIR / "aapl.csv")
amzn = pd.read_csv(DATA_DIR / "amzn.csv")
fb = pd.read_csv(DATA_DIR / "fb.csv")
goog = pd.read_csv(DATA_DIR / "goog.csv")
nflx = pd.read_csv(DATA_DIR / "nflx.csv")

print("AAPL shape:", aapl.shape)
print("AMZN shape:", amzn.shape)
print("FB shape:", fb.shape)
print("GOOG shape:", goog.shape)
print("NFLX shape:", nflx.shape)

AAPL shape: (251, 6)
AMZN shape: (251, 6)
FB shape: (251, 6)
GOOG shape: (251, 6)
NFLX shape: (251, 6)


Each source file contains 251 observations and 6 original columns. The next cell displays the first few rows of one file so that the structure can be checked.

In [16]:
aapl.head()

,date,open,high,low,close,volume
0,2018-01-02,166.9271,169.0264,166.0442,168.9872,25555934
1,2018-01-03,169.2521,171.2337,168.6929,168.9578,29517899
2,2018-01-04,169.2619,170.1742,168.8106,169.7426,22434597
3,2018-01-05,170.1448,172.0381,169.7622,171.6751,23660018
4,2018-01-08,171.0375,172.2736,170.6255,171.0375,20567766


## Part 4. Add a `ticker` Column




In [17]:
aapl["ticker"] = "AAPL"
amzn["ticker"] = "AMZN"
fb["ticker"] = "FB"
goog["ticker"] = "GOOG"
nflx["ticker"] = "NFLX"

aapl.head()

,date,open,high,low,close,volume,ticker
0,2018-01-02,166.9271,169.0264,166.0442,168.9872,25555934,AAPL
1,2018-01-03,169.2521,171.2337,168.6929,168.9578,29517899,AAPL
2,2018-01-04,169.2619,170.1742,168.8106,169.7426,22434597,AAPL
3,2018-01-05,170.1448,172.0381,169.7622,171.6751,23660018,AAPL
4,2018-01-08,171.0375,172.2736,170.6255,171.0375,20567766,AAPL


## Part 5. Combine the Five DataFrames





In [18]:
faanf = pd.concat(
    [fb, aapl, amzn, nflx, goog],
    ignore_index=True
)

print("Combined DataFrame shape:", faanf.shape)
faanf.head()

Combined DataFrame shape: (1255, 7)


,date,open,high,low,close,volume,ticker
0,2018-01-02,177.68,181.58,177.5500,181.42,18151903,FB
1,2018-01-03,181.88,184.78,181.3300,184.67,16886563,FB
2,2018-01-04,184.90,186.21,184.0996,184.33,13880896,FB
3,2018-01-05,185.59,186.90,184.9300,186.85,13574535,FB
4,2018-01-08,187.20,188.90,186.3300,188.28,17994726,FB


The combined DataFrame should have **1,255 rows and 7 columns**:

`date`, `open`, `high`, `low`, `close`, `volume`, and `ticker`.

## Part 6. Save the Combined Data as `faanf.csv`



In [19]:
faanf_path = DATA_DIR / "faanf.csv"

faanf.to_csv(faanf_path, index=False)

print("Saved file:", faanf_path)
print("File exists:", faanf_path.exists())

Saved file: /content/MSSP607/data/faanf.csv
File exists: True


## Part 7. Read `faanf.csv` and Convert Data Types

### Assignment Step 2(f)



In [20]:
stock_data = pd.read_csv(faanf_path)

stock_data["date"] = pd.to_datetime(stock_data["date"])
stock_data["volume"] = stock_data["volume"].astype("int64")

stock_data = stock_data.sort_values(
    by=["date", "ticker"]
).reset_index(drop=True)

stock_data.dtypes

,0
date,datetime64[ns]
open,float64
high,float64
low,float64
close,float64
volume,int64
ticker,object


The expected data types are:

- `date`: `datetime64[ns]`
- `volume`: `int64`

The next cell displays the first rows after sorting by date and ticker.

In [21]:
stock_data.head(10)

,date,open,high,low,close,volume,ticker
0,2018-01-02,166.9271,169.0264,166.0442,168.9872,25555934,AAPL
1,2018-01-02,1172.0000,1190.0000,1170.5100,1189.0100,2694494,AMZN
2,2018-01-02,177.6800,181.5800,177.5500,181.4200,18151903,FB
3,2018-01-02,1048.3400,1066.9400,1045.2300,1065.0000,1237564,GOOG
4,2018-01-02,196.1000,201.6500,195.4200,201.0700,10966889,NFLX
5,2018-01-03,169.2521,171.2337,168.6929,168.9578,29517899,AAPL
6,2018-01-03,1188.3000,1205.4900,1188.3000,1204.2000,3108793,AMZN
7,2018-01-03,181.8800,184.7800,181.3300,184.6700,16886563,FB
8,2018-01-03,1064.3100,1086.2900,1063.2100,1082.4800,1430170,GOOG
9,2018-01-03,202.0500,206.2100,201.5000,205.0500,8591369,NFLX


## Part 8. Find the Seven Rows with the Highest Volume



In [22]:
top_7_volume = stock_data.nlargest(7, "volume")

top_7_volume[
    ["date", "ticker", "volume"]
]

,date,ticker,volume
712,2018-07-26,FB,169803668
267,2018-03-20,FB,129851768
287,2018-03-26,FB,126116634
272,2018-03-21,FB,106598834
910,2018-09-21,AAPL,96246748
1225,2018-12-21,AAPL,95744384
1060,2018-11-02,AAPL,91328654


### Result

The seven highest-volume observations in the provided data are:

| Date | Ticker | Volume |
|---|---|---:|
| 2018-07-26 | FB | 169,803,668 |
| 2018-03-20 | FB | 129,851,768 |
| 2018-03-26 | FB | 126,116,634 |
| 2018-03-21 | FB | 106,598,834 |
| 2018-09-21 | AAPL | 96,246,748 |
| 2018-12-21 | AAPL | 95,744,384 |
| 2018-11-02 | AAPL | 91,328,654 |

## Part 9. Reshape the Data with `melt()`



In [23]:
melted_data = stock_data.melt(
    id_vars=["date", "ticker"],
    value_vars=["open", "high", "low", "close", "volume"],
    var_name="variable",
    value_name="value"
)

print("Original shape:", stock_data.shape)
print("Melted shape:", melted_data.shape)

melted_data.head(15)

Original shape: (1255, 7)
Melted shape: (6275, 4)


,date,ticker,variable,value
0,2018-01-02,AAPL,open,166.9271
1,2018-01-02,AMZN,open,1172.0000
2,2018-01-02,FB,open,177.6800
3,2018-01-02,GOOG,open,1048.3400
4,2018-01-02,NFLX,open,196.1000
5,2018-01-03,AAPL,open,169.2521
6,2018-01-03,AMZN,open,1188.3000
7,2018-01-03,FB,open,181.8800
8,2018-01-03,GOOG,open,1064.3100
9,2018-01-03,NFLX,open,202.0500


The original combined data contain **1,255 rows**.

There are five measurement columns being melted, so the long-format DataFrame contains:

**1,255 × 5 = 6,275 rows**

Therefore, the expected melted shape is **6,275 rows × 4 columns**.

## Part 10. Final Verification


In [24]:
assert stock_data.shape == (1255, 7)
assert pd.api.types.is_datetime64_any_dtype(stock_data["date"])
assert pd.api.types.is_integer_dtype(stock_data["volume"])
assert set(stock_data["ticker"].unique()) == {
    "AAPL", "AMZN", "FB", "GOOG", "NFLX"
}
assert melted_data.shape == (6275, 4)
assert faanf_path.exists()

print("All assignment checks passed.")
print("Combined rows:", len(stock_data))
print("Tickers:", sorted(stock_data["ticker"].unique()))
print("Melted rows:", len(melted_data))
print("faanf.csv saved:", faanf_path.exists())

All assignment checks passed.
Combined rows: 1255
Tickers: ['AAPL', 'AMZN', 'FB', 'GOOG', 'NFLX']
Melted rows: 6275
faanf.csv saved: True
